README.md
# contamMix
Contamination estimation from mitochondrial ancient DNA

This repository contains code to generate a R package ("contamMix") that implements a mixture-model used to estimation contamination in ancient  mitochondrial DNA using MCMC.  The model is detailed in supplementary information 5 ("Likelihood-based method for contamination estimation") of the following paper, which remains the best citation if using this method in a publication:

> Fu Q, Mittnik A, Johnson PLF, Bos K, Lari M, Bollongino R, Sun C, Giemsch L, Schmitz R, Burger J, Ronchitelli AM, Martini F, Cremonesi RG, Svoboda J, Bauer P, Caramelli D, Castellano S, Reich D, Pääbo S, and Krause J. (2013). A revised timescale for human evolution based on ancient mitochondrial genomes. *Curr Biol*, 23:553-9. [[doi]](http://dx.doi.org/10.1016/j.cub.2013.02.044)

### Source code context

Code was maintained in a svn repository and shared upon request until Sept 2022, when it was imported into this public GitHub repository.  The R code calls a perl script to pre-process the data (exec/sam2mn.pl) and then calls a C entrypoint to a Gibbs sampler written in C++ (src/contam_gibbs.cpp).  R itself is only used for visualization and to hold the other code together.

### Usage

From the directory above this repository, you can build a R package with the command

	R CMD build contamMix

The resulting R "source" package then needs to be installed into your R installation (note the filename generated by the above command will depends on the version in the DESCRIPTION file):

	R CMD INSTALL contamMix_1.0-10.tar.gz

Contamination estimation:
   1) Generate a FASTA-formatted multiple alignment of potential
contaminants + authentic (consensus) sequence.  Two possible programs for
making this alignment are mafft and muscle.

   2) Align your reads to your inferred consensus sequence, perhaps
using bwa.

   3) Run "estimate.R" (included in the contamMix package under the
"exec" subdirectory, so installed wherever the R package was installed). Ex:

	/usr/local/lib/R/site-library/contamMix/exec/estimate.R --samFn data.bam --malnFn aln.fa --figure data_fig

If you run "estimate.R" without any options, it will display a usage
summary.  Under the hood, estimate.R really just calls four functions
from the package: loadSAM, runChains, print.contamMix, and
plot.contamMix. You can also use these functions directly; the package
includes help pages for each.

## Caveats:
The method's key assumption is that contamination is <50%, and thus that the
consensus reflects the authentic sequence.  The consensus sequence
should not have ambiguity characters at positions that are variable in
your data -- contamMix ignores sites with ambiguity characters, so this
would be effectively throwing out your informative sites .  As always
with MCMC, achieving convergence can be fiddly.  I suggest always
graphically checking for convergence in addition to looking at the
Gelman-Rubin diagnostic.  For some datasets, I have had to tweak the
"alpha" hyperparameter (this is an option to estimate.R).


NEWS
v1.0-11 -- 26 Sep 2022
  - explicit calls to contamMix S3 class functions to avoid errors in certain newer versions of R.
  - more flexible multithreading, for compatibility with HPCs.

v1.0-10 -- 16 Jan 2014
  - add trivial configure script to twiddle shebang in exec/estimate.R

v1.0-9 -- 14 Nov 2013
  - rename --save option to estimate.R to avoid accidental partial
    matching to --saveMN

v1.0-8 -- 13 Nov 2013
  - added option to save MN intermediate file.
  - warn to stderr on any read at least 30bp long with >50% mismatches
  - bug fix: regression introduced in 1.0-7 with R contamMix class not
    being maintained

v1.0-7 -- 6 Nov 2013
  - added option to trim a given number of bases from the ends of each read (another way of avoiding problems from deamination).

v1.0-6 -- 29 Oct 2013
  - bug fix: not marking indels in multiple align if aligned to ambiguity character (e.g., N)

v1.0-5 -- 15 Jul 2013
  - issue error if potential contamination sequences appear unaligned (i.e. of different lengths)
  - expose --transverOnly option for restricting to transversion sites

v1.0-4 -- 26 Jun 2013
  - display estimated error rate in figure output

v1.0-3 -- 18 Jun 2013
  - halt execution if estimate error rate is >15% -- this is
    implausibly high and likely signals an alignment error.

v1.0-2 -- 12 Jun 2013
  - add --baseq parameter to estimate.R
  - more informative error message if using -baseq and BAM has no q scores
  - better error messages if BAM mapping is invalid

v1.0-1 -- 7 Jun 2013
  - Bug fix in estimate.R (had some incorrect parameter names)
  - More informative i/o error messages in sam2mn.pl
  - add --consId parameter to estimate.R (if name used as BAM reference differs from that used in alignment)

R/load.R
loadSAM <- function(samFn, malnFn, endogId = NULL, baseqThreshold = 30,
                    transverOnly = FALSE, trimBases = 0, saveMN = FALSE) {
  ## find our perl script
  sam2mn = system.file("exec/sam2mn.pl", package="contamMix")
  if (sam2mn == "") stop("could not find 'sam2mn.pl' script!")

  ## error checks
  if (is.null(samFn)  ||  !file.exists(samFn))
    stop("Cannot find SAM file '", samFn, "'.")
  if (is.null(malnFn)  ||  !file.exists(as.character(malnFn)))
    stop("Cannot find multiple alignment file '", malnFn, "'.")
  
  ## load data
  optnotnull <- function(opt,val) (
    if (is.null(val)  ||  val == ""  ||  val == FALSE)
    "" else paste(opt,val))
  cmd = paste("perl -w", sam2mn, "-bam", samFn, "-fa", malnFn,
      optnotnull("-ref", endogId), optnotnull("-baseq", baseqThreshold),
      optnotnull("-transverOnly", transverOnly),
      optnotnull("-trimBases", trimBases))
  if (!saveMN) {
      x = read.table(pipe(cmd), comment.char='', fill=TRUE)
  } else {
      mnFilename = paste(samFn,".mn", sep='');
      system(paste(cmd, ">", mnFilename))
      x = read.table(mnFilename, comment.char='', fill=TRUE)
  }

  data=x_loadMNtable(x)
  data$samFn=samFn
  data
}

x_loadMNtable <- function(x) {
  ## parse out error estimate
  if (x[nrow(x),1] != "#error:") { # independently estimated via other means
    stop("error parsing input data")
  }
  priorErrorRate = as.numeric(as.character(x[nrow(x),2]))
  if (priorErrorRate == 0) { priorErrorRate = 1e-6 } #avoid log(0)s
  if (priorErrorRate > 0.15) {
    stop("estimated error rate is implausibly high -- check for error in alignments")
  }

  ## reshape data into tensor (3-d matrix)
  numMT = (ncol(x)-1)/2
  mnMatrix = as.matrix(x[-nrow(x),-1])
  dim(mnMatrix) = c(nrow(mnMatrix), 2, numMT)
  dimnames(mnMatrix) = list(x[-nrow(x),1], c("M","N"), 0:(numMT-1))

  structure(list(numMT=numMT, mnMatrix=mnMatrix, e=priorErrorRate),
            class = "contamMix")
}


R/output.R
## ---------------------------------------------------------------------------
## pre : x and y coords intended for plotting, desired output dpi
## post: which coords should be plotted for optimal display at specified dpi
x_thinXY <- function (x, y=NULL, type='p', dpi=20, xlim=NULL, ylim=NULL) {
  xy = xy.coords(x, y)
  if (is.null(xlim)) xlim = range(xy$x)
  if (is.null(ylim)) ylim = range(xy$y)
  uin = c(diff(xlim), diff(ylim))/par("pin") #usr coord / inch
  udt = uin / dpi #usr coord / dot
  dtx = round(xy$x / udt[1])
  dty = round(xy$y / udt[2])
  if (type == 'p') {## if plotting points, just eliminate all dups
    w=!duplicated(cbind(dtx,dty))
  } else if (type == 'l') { ## if plotting lines, eliminate all *pairs* of dups
    n = length(dtx)
    z=cbind(dtx[-n],dty[-n],dtx[-1],dty[-1])
    ## line is same either direction
    rv = z[,1] > z[,3]  |  (z[,1] == z[,3]  &  z[,2] > z[,4])
    z[rv,] = z[rv,c(3,4,1,2)]
    gdPairs = !duplicated(z)  &  (dtx[-n] != dtx[-1]  |  dty[-n] != dty[-1])
    w=c(gdPairs, FALSE)  |  c(FALSE, gdPairs)
  }
  xy$x = xy$x[w]
  xy$y = xy$y[w]
  return(xy)
}

x_AnalyzeChains <- function(chains, focalChain=1) {
  ## crude identification of burn-in
  ## could just use "lik", but empirical results suggest "0" is often better..
  burnIn = max(sapply(chains, function(x) (max(
    which(x[,"0"]   > median(x[(nrow(x) %/% 2):nrow(x),"0"])  )[1],
    which(x[,"lik"] > median(x[(nrow(x) %/% 2):nrow(x),"lik"]))[1]))
    ))

  if (is.na(burnIn)) {
    warning("MCMC not run long enough to guestimate burn in; proceeding using no burn in for diagonostic purposes.")
    burnIn=1;
  }

  ## arbitrarily designate one run to analyze
  chain = chains[[focalChain]]
  samples = chain[-(1:burnIn),"0"]

  ## estimate posterior density after burn-in
  d=density(samples, cut=0);
  mcmcObj = coda::mcmc.list(lapply(chains, function(x) (coda::mcmc(x[,"0"]))))

  list(burnIn=burnIn, chain=chain, samples=samples, dens=d,
       MAP = d$x[which.max(d$y)], ci=quantile(samples, c(0.025, 0.975)),
       mcmcObj=mcmcObj)
}

## ---------------------------------------------------------------------------
## Public functions below

print.contamMix <- function(x, tabDelim = FALSE, ...) {
  if (!tabDelim) {
    cat("Data from:\n  ", x$samFn, "\nconsist of", nrow(x$mnMatrix), "reads,",
        "considered as a mixture of", x$numMT, "possible genomes.\n")

    contamEvidence = apply(x$mnMatrix, 1, function(d) (any(d["M",1]<d["M",-1])))
    cat("Crude contamination upper bound: ", sum(contamEvidence), " out of ",
        nrow(x$mnMatrix), " reads (",
        signif(100*sum(contamEvidence)/nrow(x$mnMatrix), 2),
        "%) match another\n",
        "genome better than the consensus (error rate is ",
        x$e, ").\n\n", sep='')

    if (is.null(x$chains)) {
      cat("[mcmc estimate not available since chain hasn't been run]\n\n")
    }
  }
  
  if (is.null(x$chains)) return(invisible())

  ## arbitrarly focus on one chain
  anal = x_AnalyzeChains(x$chains, focalChain=1)

  if (tabDelim) {
    cat(x$e, anal$MAP, anal$ci,
        if (length(x$chains) > 1) coda::gelman.diag(anal$mcmcObj)$psrf else "",
        sep="\t");
    cat("\n")
  } else {
    cat("quantiles from n = ", length(anal$samples),
        " samples (after discarding burnin of ", anal$burnIn, ")\n", sep='')
    print(anal$ci)
    cat("MAP authentic:", anal$MAP, "\n");
    if (length(x$chains) > 1) print(coda::gelman.diag(anal$mcmcObj))
  }
  
  invisible()
}


plot.contamMix <- function(x, which=1:3, ...) {
  if (is.null(x$chains)) stop("MCMC not yet run (use runChain)")
  
  anal = x_AnalyzeChains(x$chains, focalChain=1)
  
  ##convergence diagnostic, if applicable
  if (1 %in% which  &&  length(x$chains) > 1) {
    coda::gelman.plot(anal$mcmcObj, auto.layout=FALSE, ylim=c(1,2))
  }

  if (2 %in% which) {
    plot(x_thinXY(anal$chain[,"0"], dpi=60), type='l', xlab='Iteration', ylab='Pr(authentic)')
    abline(v=anal$burnIn, col = "blue", lty=3)
  }

  if (3 %in% which) {
    plot(anal$dens, type='n', main=NA, xlab="Pr(authentic)", xlim=c(0,1))
    ciRegion = anal$dens$x >= anal$ci[1] & anal$dens$x <= anal$ci[2]
    polygon(cbind(c(anal$ci[1], anal$dens$x[ciRegion], anal$ci[2]),
                  c(0, anal$dens$y[ciRegion], 0)), border=NA, col='gray')
    lines(anal$dens) #plot density on top of shaded region
    abline(v=anal$MAP, lty=2, col='red')
    text(x=anal$MAP, y=mean(par()$usr[3:4]), labels=
         paste("~", format(anal$MAP, nsmall=2, digits=2), sep=''),
         col='red', cex=1.2, adj=c(1.2, .5), xpd=NA)
  }
  
  invisible()
}


R/runChains.R
x_gibbs.step <- function(D, prop, alpha) {
  .Call("gibbs_step", D, prop, alpha)
}

## rdirichlet code posted by Ben Bolker to R-News on Fri Dec 15
## 2000. See http://tolstoy.newcastle.edu.au/R/help/00b/2175.html. Ben
## attributed the code to Ian Wilson i.wilson@maths.abdn.ac.uk.
x_rdirichlet <- function (n, alpha) 
{
    l <- length(alpha)
    x <- matrix(rgamma(l * n, alpha), ncol = l, byrow = TRUE)
    sm <- x %*% rep(1, l)
    return(x/as.vector(sm))
}

x_runChain <- function(y, nIter, alpha, recordAll) {
  res = matrix(0, nrow=nIter, ncol=1 + ifelse(recordAll, ncol(y),1))
  colnames(res) = c("lik", 0:(ncol(res)-2));
  ## initialize chain at random proportions
  mcmcProp = as.vector(x_rdirichlet(1, rep(alpha, ncol(y))))
  ## iterate MCMC
  for (i in 1:nIter) {
    res[i,]=c(sum(log(y %*% mcmcProp)),
         if (recordAll) mcmcProp else mcmcProp[1])
    mcmcProp = x_gibbs.step(y, mcmcProp, alpha)
  }
  return(res)
}

runChains <- function(data, nChains, nIter, alpha = 0.1, recordAll = FALSE) {
  if (class(data) != "contamMix") stop("data is of wrong type")
  if (nChains < 1) stop("nChains must be a positive number")
  if (nIter < 1) stop("nIter must be a positive number")
  if (alpha < 0) stop("alpha must be non-negative number")

  ## incorporate errors into data
  y = (1-data$e)^data$mnMatrix[,"M",] * data$e^data$mnMatrix[,"N",]
  if (is.vector(y)) { y = t(y) } #edge case if only one read

  ## run multiple chains, ideally in parallel
  data$chains =
    parallel::mclapply(1:nChains, function(i) (
      x_runChain(y, nIter, alpha, recordAll) ))
  data
}


exec/sam2mn.pl
#!/usr/bin/perl -w

# Copyright 2012,2013 Philip Johnson.
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published
#     by the Free Software Foundation, either version 3 of the License,
#     or (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License at <http://www.gnu.org/licenses/> for
#     more details.

# Based off earlier (more general) sam2mtdiff.pl + implement baseQ thresholding
#
# Inputs: take BAM/SAM alignment against consensus, fasta-formatted
# multiple alignment (mafft?) of full-length mtDNA genomes (including consensus), id for consensus in multiple alignment (by default uses reference id in BAM).
#
# Output: one line per read with the following columns:
#  <read id>
#  <read len>
#  <# matches to consensus>
#  <# non-matches to consensus>
#  <# matches to genome 1>
#  <# non-matches to genome 1>
#  ...
#  <# matches to genome n>
#  <# non-matches to genome n>
#
# And the last line contains an estimate of error

use strict;
use Getopt::Long;

use constant { #name SAM fields
    ID => 0,
    FLAG => 1,
    REFID => 2,
    POS => 3,
    MAPQ => 4,
    CIGAR => 5,
    MRNM => 6,
    MPOS => 7,
    ISIZE => 8,
    SEQ => 9,
    QUAL => 10,
    TAG => 11
};

my ($bamFN, $contamFN, $consId, $baseQthreshold);
my $mapQthreshold = 30;
my $transverOnly;
my $trimBases = 0; # number of bases to trim off each end of read (simple hack to minimize aDNA damage)
my $warnId = 1; #warn if reference id does not match supplied consId
if (!GetOptions('bam=s' => \$bamFN,
                'ref=s' => \$consId,
                'fa=s' => \$contamFN,
                'mapq=i' => \$mapQthreshold,
                'baseq=i'=> \$baseQthreshold,
                'trimBases=i' => \$trimBases,
                'transverOnly'=> \$transverOnly) ||
    !defined($bamFN  &&  $contamFN)) {
    print "usage: $0 -bam <adna.bam> -fa <genomes.fa> [-ref <consensus id>] [...]\n".
        "where ... could be:\n".
        "\t-mapq <30>  minimum map quality (defaults to 30)\n".
        "\t-baseq <>  minimum base quality (defaults to no filter)\n".
        "\t-transverOnly  (only count transversions)\n".
        "\t-trimBases <0>  num bases to trim/ignore at ends of each read (think aDNA damage)\n".
        "\n";
    exit 1;
}

## load all potential contaminate mt genomes into memory (311 * 16e3 = ~5MB -- no prob)
my (@genomes, $alignedLength, $alignedConsensus);
{
    my ($acc, $seq);
    open INFA, $contamFN or die "$! attempting to open '$contamFN'\n";
    while (<INFA>) {
        if (/^>/) {
            push @genomes, [$acc, $seq] if defined $seq;
            ($acc) = />(.+?)\s/;
            if ($acc =~ /|/) { # try to pick out actual accession
                foreach my $id (split /\|/, $acc) {
                    $acc = $id if length($id)>2  &&  $id =~ /^\w/;
                }
            }
            $seq = [];
        } else {
            chomp;
            $_ = uc $_;
            tr/ACGTN-/N/c; #restrict to ACGTN-
            push @$seq, split //, $_;
        }
    }
    push @genomes, [$acc, $seq] if defined $seq;
    close INFA;
    
    $alignedLength = scalar(@{$genomes[0][1]});
}

# extra check that genomes are properly multiply-aligned
for (my $j = 1;  $j < @genomes;  ++$j) {
    if ($alignedLength != scalar(@{$genomes[$j][1]})) {
        die "authentic + potential contaminant genomes do not appear to be aligned -- they are different lengths!\n"
    }
}


# propogate a N in any genome to all ungapped genomes & note status of
# following at positions across all genomes:
my @fixedPositions = (0) x $alignedLength;
my @transverPositions = (0) x $alignedLength;
my @tiPositions = (0) x $alignedLength;
my @indelPositions = (0) x $alignedLength;
for (my $i = 0;  $i < $alignedLength;  ++$i) {
    my $foundN = 0;
    my %bases = ($genomes[0]->[1][$i] => 1);
    for (my $j = 0;  $j < @genomes;  ++$j) {
        if ($genomes[$j]->[1][$i] eq 'N') {
            $foundN = 1;
            last;
        } elsif ($genomes[$j]->[1][$i] ne $genomes[0]->[1][$i]) {
            $bases{$genomes[$j]->[1][$i]} = 1;
        }
    }
    if ($foundN) {
        for (my $j = 0;  $j < @genomes;  ++$j) {
            if ($genomes[$j]->[1][$i] ne '-') {
                $genomes[$j]->[1][$i] = 'N';
            } else {
                $indelPositions[$i] = 1;
            }
        }
    } else {
        if (exists($bases{'-'})) {
            $indelPositions[$i] = 1;
        } elsif (keys(%bases) == 1) {
            $fixedPositions[$i] = 1;
        } elsif (keys(%bases) == 2  &&
                 !((exists($bases{'C'})  &&  exists($bases{'T'}))  ||
                   (exists($bases{'G'})  &&  exists($bases{'A'})))) {
            $transverPositions[$i] = 1;
        } elsif (keys(%bases) == 2) {
            $tiPositions[$i] = 1;
        }
    }
}

# print "#fixed : ", scalar(grep {$_} @fixedPositions), "\n";
# print "#transv: ", scalar(grep {$_} @transverPositions), "\n";
# print "#indel : ", scalar(grep {$_} @indelPositions), "\n";
# print "#ti    : ", scalar(grep {$_} @tiPositions), "\n";
# print "#other : ", @fixedPositions-(scalar(grep {$_} @indelPositions)+scalar(grep {$_} @transverPositions)+scalar(grep {$_} @fixedPositions)+scalar(grep {$_} @tiPositions)), "\n";


my @alnMap; # for mapping between alignment & consensus coordinates


## iterate through all reads and compare to each potential contaminate genome
my $inferringConsensus = 0; #whether we are inferring the consensus id
my ($errDenom, $errNumer) = (0,0); #for simple estimate of errors
open INBAM, ($bamFN =~ /.sam$/) ? $bamFN : "samtools view $bamFN |" or
    die "$! attempting to open '$bamFN'\n";
while (<INBAM>) {
    my @F = split /\t/;
    next if ($F[MAPQ] < $mapQthreshold  ||  $F[FLAG] & 0x4); #"unmapped" flag
    --$F[POS]; #convert to 0-based coords;

    # on first line, identify consensus and initialize mapping
    if (@alnMap == 0) {
        if ($warnId  &&  defined($consId)  &&  $F[REFID] ne $consId) {
            print STDERR "Warning: BAM reference id ('$F[REFID]') does not match supplied consensus id ('$consId')\n";
        } else {
            $consId = $F[REFID];
            $inferringConsensus = 1;
        }
        
        my $consI;
        for ($consI = 0;  $consI < @genomes  &&  $genomes[$consI][0] ne $consId;
             ++$consI) {};
        if ($consI == @genomes) {
            if ($inferringConsensus) {
                die "Error: SAM/BAM reference id ('$consId') does not exist in the multiple alignment.  Override with 'consId' command line option if necessary -- but ensure both identifiers do truly correspond to the consensus sequence!\n";
            } else {
                die "Error: Couldn't find user-supplied consensus '$consId' in multiple alignment!\n";
            }
        }
        $alignedConsensus = $genomes[$consI][1];
        # ensure consensus at first position in array
        unshift(@genomes, $genomes[$consI]);
        splice(@genomes, $consI+1, 1);

        # make mapping between consensus and alignment coordinates
        my $seqI = 0;
        for (my $alnI = 0;  $alnI < $alignedLength;  ++$alnI) {
            if ($alignedConsensus->[$alnI] ne '-') {
                $alnMap[$seqI++] = $alnI;
            }
        }
    }


    # Expand read into @genomes alignment space using CIGAR string
    # (which gives read::consensus mapping) and @alnMap (which gives
    # consensus::alignment mapping)
    #print "cigar: ", $F[CIGAR], "\n";
    my @bases;
    my $discard = 0;
    my ($rPos, $sPos) = (0, $F[POS]);
    while ($F[CIGAR] =~ /(\d+)([SMDI])/g) {
        if ($2 eq 'S') { #soft clipping
            $rPos += $1;
        } elsif ($2 eq 'M') { #match
            die "error processing read '$F[ID]' -- mapped position (".($sPos+1).") is off end of reference sequence (length ".scalar(@alnMap).")!\n" if ($sPos >= @alnMap);
            my $aPos = $alnMap[$sPos];
            my @match = split //, substr $F[SEQ], $rPos, $1;
            if (defined($baseQthreshold)) { #"N"-out bases below threshold
                if ($F[QUAL] eq '*') {
                    print STDERR "Warning: cannot use base Q threshold since sam/bam does not include base qualities for read '$F[ID]'\n";
                } else {
                    my @quals = split //, substr $F[QUAL], $rPos, $1;
                    for (my $k = 0;  $k < @quals;  ++$k) {
                        if (ord($quals[$k])-33 < $baseQthreshold) {
                            $match[$k] = 'N';
                        }
                    }
                }
            }
            
            if ($aPos + $1 > @$alignedConsensus) {
                warn "warning: read '$F[ID]' mapping is partially off end of reference sequence (discarding read)!\n" ;
                $discard = 1;
                last;
            } else {
                for (my $i = 0;  $i < $1;  ++$i) {
                    while ($alignedConsensus->[$aPos++] eq '-') {
                        push @bases, '-';
                    }
                    push @bases, $match[$i];
                }
                $rPos += $1;
                $sPos += $1;
            }
        } else { #indel -- discard due to potential flaky alignment
            $discard = 1;
            last;
        }
    }

    #print $alnMap[$F[POS]], "\t", $alnMap[$F[POS]]+@bases-1, "\n";
    #print @bases, "\n";
    #print @{$alignedConsensus}[($alnMap[$F[POS]])..($alnMap[$F[POS]]+@bases-1)], "\n";
    #print "----------\n";
    #if requested, trim bases from each end of read
    my $alnStart = $alnMap[$F[POS]];
    my $aLeft = 0;
    for (my $sI=0;  $aLeft < @bases  &&  $sI < $trimBases;  ++$aLeft) {
        ++$sI if $bases[$aLeft] ne '-';
    }
    my $aRight = @bases-1;
    for (my $sI=0;  $aRight >= 0  &&  $sI < $trimBases;  --$aRight) {
        ++$sI if $bases[$aRight] ne '-';
    }
    if ($aLeft == @bases  ||  $aRight < 0) {
        $discard = 1;
    } else {
        $alnStart += $aLeft;
        @bases = splice @bases, $aLeft, $aRight-$aLeft+1;
    }

    # finally summarize read!
    my @readSummary;
    if (!$discard) {
        for (my $j = 0;  $j < @genomes;  ++$j) {
            my ($M,$N,$I) = (0,0,0);
            for (my ($k, $alnPos) = (0,$alnStart);  $k < @bases;
                 ++$k, ++$alnPos) {
                if ($indelPositions[$alnPos]) {
                    $discard = 1;
                    last;
                }
                my $genomeB = $genomes[$j][1][$alnPos];
                next if (defined $transverOnly  &&
                         !$transverPositions[$alnPos]);
                if ($bases[$k] eq $genomeB) {
                    ++$M if ($genomeB ne 'N') #only count match if not N-to-N
                } else {
                    ++$N if ($bases[$k] ne 'N'  &&  $genomeB ne 'N');
                }
            }
            push @readSummary, [$M, $N];
        }
    }
    if ($discard) {
        #print STDERR "fyi: discarding read $F[ID]\n";
        next;
    }

    # simple estimate of error rate
    my $allFixed = 1;
    for (my $k = 0;  $k < @bases;  ++$k) {
        my $alnPos = $k + $alnStart;
        if ($fixedPositions[$alnPos]) {
            my $genomeB = $genomes[0][1][$alnPos];
            if ($bases[$k] ne 'N'  &&  $genomeB ne 'N') {
                ++$errDenom;
                ++$errNumer if ($bases[$k] ne $genomeB)
            }
        } else {
            $allFixed = 0;
        }
    }
    
    # report read only if informative
    if (!$allFixed  &&  $readSummary[0][0] + $readSummary[0][1] > 0) {
        print $F[ID], "\t";
        for (my $j = 0;  $j < @genomes;  ++$j) {
            print join(' ', @{$readSummary[$j]}), ' ';
        }
        print "\n"
    } elsif (($readSummary[0][0] + $readSummary[0][1]) > 30  &&
        $readSummary[0][1]/($readSummary[0][0] + $readSummary[0][1]) > 0.50) {
        warn("implausibly high number of mismatches in read '$F[ID]'\n");
    }
}
close INBAM;


print "#error:\t";
if ($errDenom > 0) {
    print sprintf('%.4f', $errNumer/$errDenom), "\t$errDenom";
} else { 
    print "0\t0";
}
print "\n";


exec/estimate.R
library(contamMix)
library(getopt)
# options(mc.cores = parallel::detectCores()) ## Original behaviour of using all available cores.

## load arguments from command line (if any) & set defaults
opt = getopt(rbind(
  c('samFn', NA, 1, "character"),
  c('malnFn', NA, 1, "character"),
  c('consId', NA, 1, "character"),
  c('figure', NA, 1, "character"), #if !interactive, save figure to this file
  c('nIter', NA, 1, "integer"),
  c('nChains', NA, 1, "integer"),
  c('alpha', NA, 1, "double"),
  c('baseq', NA, 1, "integer"),
  c('transverOnly', NA, 0, "logical"),
  c('trimBases', NA, 1, "integer"),
  c('saveData', NA, 1, "character"),#save MCMC data for manual diag
  c('saveMN', NA, 0, "logical"),    #save MN intermediate for debugging
  c('recordAll', NA, 0, "logical"), #record *all* proportions from MCMC chain
  c('tabOutput', NA, 0, "logical"), #simple tab-delimited output
  c('nrThreads', NA, 1, "integer")) # Number of threads to use. Defaults to 1. 
  )
if (is.null(opt$nIter))  opt$nIter = 50000
if (is.null(opt$nChains)) opt$nChains = 3
if (is.null(opt$alpha)) opt$alpha = 0.1
if (is.null(opt$baseq)) opt$baseq = 30; if (opt$baseq == 0) opt$baseq = NULL
if (is.null(opt$recordAll)) opt$recordAll = FALSE
if (is.null(opt$saveMN)) opt$saveMN = FALSE
if (is.null(opt$tabOutput)) opt$tabOutput = FALSE
if (is.null(opt$nrThreads)) opt$nrThreads = 1

if (!interactive()  &&  (is.null(opt$samFn)  ||  is.null(opt$malnFn))) {
  cat("Usage: ./estimate.R --samFn <data.sam>  --malnFn <alignment.fa> [--nIter <50000>] [--nChains <3>] [--alpha <0.1>] [--figure <outputfigure.pdf>] [--baseq <30>] [--transverOnly] [--trimBases <0>] [--saveMN] [--tabOutput] ...\n\n",
      "Required parameter:\n",
      "\t--samFn    --> SAM/BAM data file aligned to consensus\n",
      "\t--malnFn   --> FASTA-format multiple alignment of consensus and potential contaminants\n",
      "Optional parameters:\n",
      "\t--consId --> consensus id in multiple alignment (if not supplied, assumes this is identical to the reference id in the SAM/BAM)\n",
      "\t--nIter --> number of iterations in Markov chain (default 50000)\n",
      "\t--nChains --> number of MCMC chains to run from different random starting parameters, which are used for Gelman-Rubin convergence testing (default 3; uses multiple processors if available)\n",
      "\t--alpha --> hyperparameter to Dirichlet prior distribution; may need to be tweaked if MCMC is mixing poorly (default 0.1)\n",
      "\t--figure --> if supplied, generates a PDF figure with 3 panels: convergence of gelman diagnostic (if --nChains>1),  Pr(authentic) as a function of MC iteration, estimated posterior density for Pr(authentic)\n",
      "\t--baseq --> base quality threshold below which to discard data (default 30; 0 signals NO threshold)\n",
      "\t--transverOnly --> only use sites with transversions (avoids potential for bias from aDNA damage, but has significantly less power)\n",
      "\t--trimBases --> trim this # of bases from ends of sequence (reducing effect of aDNA damage)\n",
      "\t--saveData --> save chain data to specified file (in .Rdata format) for manual diagnostics\n",
      "\t--saveMN --> save MN intermediate file for manual debugging (will use filename '<samFn>.mn')\n",
      "\t--nrThreads --> The number of threads to use. Defaults to 1. If a number higher than the available threads is used, the maximum available number of threads will be used instead.\n",
      "\t--tabOutput --> output a single line of text with the following tab-separated values: <inferred-error-rate> <MAP-authentic> <2.5% authentic> <97.5% authentic> <gelman diagnostic> <gelman diag upper bound>\n\n")
  cat("NOTE: Not converged if Gelman diagnostic >1.1.  Initial runs on any new data should request a --figure to visually check that the Markov chain is well-behaved.\n\n")
  q()
}

## Use minimum of max available cores and given threads. Avoids overspecifying of cores, and hogging an entire HPC.
used_cores = min(parallel::detectCores(), opt$nrThreads)
options(mc.cores = used_cores)
cat(paste0("Using ",used_cores, " threads.\n"), file=stderr())

## test only
#data = loadSAM(samFn="~/data/emh/mt_only/Paglicci23_sequence_merged_quality_Monly-human_MT.sort.q30_l35_uniq.bam", malnFn="~/data/emh/mt_only/aln.mafft", opt$alnFn, endogId="Paglicci23"

## do the work
data = loadSAM(samFn=opt$samFn, malnFn=opt$malnFn, endogId=opt$consId, baseqThreshold=opt$baseq, transverOnly=opt$transverOnly, trimBases=opt$trimBases, saveMN=opt$saveMN)
res = runChains(nChains=opt$nChains, nIter=opt$nIter, data=data, alpha=opt$alpha)

## results in text format
print.contamMix(res, tabDelim=opt$tabOutput)

## binary save of results
if (!is.null(opt$saveData)) {
  if (!grepl(".Rdata$", opt$saveData))
      opt$saveData = paste(opt$saveData, ".Rdata", sep="")
  if (!opt$tabOutput) cat("Saving data & results to", opt$saveData, "\n")
  save(res, file=opt$saveData)
}

## graphical results
if (!is.null(opt$figure)) {
  if (!grepl(".pdf$", opt$figure)) opt$figure = paste(opt$figure, ".pdf", sep="")
  if (!opt$tabOutput) cat("Saving figure to", opt$figure, "\n")
  
  dev.new(file=opt$figure, width=4, height=6, pointsize=10)
  par(mfrow=c(3,1), cex=1, mex=0.5, mar=c(5,5,1,1), oma=c(0,0,3,0))
  plot.contamMix(res, which=1:3)
  mtext(opt$samFn, side=3, line=1, outer=TRUE, font=2)
  mtext(bquote(paste("(", epsilon, " = ", .(signif(res$e,2)), ")")), side=3, line=-1, outer=TRUE)
  invisible(dev.off())
}
